### PR TITLE
Add opt-in Claude Code mode classifier

### LINKED ISSUE #274: Feature: opt-in install flag to replace PAI ModeClassifier with soma's classifier (claude-code)
## Goal
Let `soma install claude-code` optionally install a UserPromptSubmit hook that drives Claude's mode/effort decision from **soma's own classifier** (`src/algorithm-classifier.ts`) instead of PAI's `ModeClassifier.hook.ts`. Unifies mode classification across substrates (codex/pi already use soma's classifier) and adds effort-tier awareness on the Claude hop.

## Why
- **Unification**: one classifier brain across Claude / codex / pi.dev instead of two parallel impls (PAI `ModeClassifier.hook.ts` is deterministic binary minimal-vs-algorithm; soma's is deterministic 3-mode + E1–E5 effort).
- **Effort-tier awareness** on Claude (PAI's hook emits only MODE, no tier).

## Scope
- **A. Hook asset** — new claude-code `UserPromptSubmit` hook that shells `soma algorithm classify --prompt <stdin> --json`, maps `{mode,effort}` → the executor additionalContext (MODE header + the no-downshift instruction). Mirror the existing `soma-claude-code-hook` render/install pattern.
- **B. Install flag** — `soma install claude-code --mode-classifier` (or config toggle) writes the hook + its `settings.json` UserPromptSubmit entry **and disables PAI's `ModeClassifier` entry** (two UserPromptSubmit classifiers conflict — both emit MODE). `uninstall` restores PAI's.
- **C. Classifier parity (the real work)** — soma's minimal/native detection is narrower than PAI's. Empirically `soma algorithm classify` returns `mode:algorithm` for `"ok thanks"` and `"what time is it"` (PAI's EN+RU greeting/ack regex catches those → MINIMAL). Port PAI's minimal/native patterns + tune NATIVE_PATTERNS so we don't trade one bug for an over-ceremony regression. Needs unit tests for the minimal/native/algorithm + effort boundaries.

## Non-goal / caveat
This does NOT fix the OBSERVE downshift bug (executor bailing to NATIVE on complex tasks) — that was PAI `ModeClassifier.hook.ts` line 159's downshift license, already fixed PAI-side by tightening it to 'downshift ONLY for trivial single-step; multi-deliverable/design/build MUST run all 7 phases'. The soma hook must emit the SAME tightened no-downshift instruction.

## Provenance note
`ModeClassifier.hook.ts` is PAI (lives in the user's ivy-shared `~/.claude/hooks/`, scaffolded from PAI). soma neither owns nor projects it today; this feature would let soma optionally take over that slot.

## Open Qs
- Pass soma's `effort` (E1–E5) through to the executor as a TIER hint?
- Default off (opt-in) vs eventually default-on once parity (C) is proven via tests.

### PR DESCRIPTION
**Pull request by jcfischer**
Closes #274.

## Summary

- adds `soma install claude-code --mode-classifier` as an opt-in install flag
- installs a Claude Code `UserPromptSubmit` hook that calls `soma algorithm classify --json` and injects concise MODE context
- disables detected PAI `ModeClassifier.hook.*` entries during opt-in install and restores them on uninstall
- tunes Soma classification for minimal acknowledgements and simple native time/date questions

## Verification

- `rtk bun run typecheck`
- `rtk bun run lint`
- `rtk bun test test/algorithm.test.ts`
- `rtk bun test test/claude-code-install.test.ts`
- `rtk bun test test/cli.test.ts`
- `rtk bun test`

### CODE REVIEW

**Review by jcfischer — COMMENTED** [review]
## Sage code review — changes-requested

3 finding(s): 2 important, 1 suggestion.


### CodeQuality
The change is mostly straightforward for the CodeQuality lens, but one CLI parsing edge case breaks valid installs. The classifier hook and settings patching paths are otherwise covered by focused tests in the diff.

- **[important]** `src/cli/substrate-lifecycle.ts:241` — **Flag order rejects valid installs**
  The parser rejects `--mode-classifier` before it has consumed a later substrate override because it checks `if (substrate !== "claude-code")` against the initial positional substrate. A command like `soma install codex --mode-classifier --substrate-home ...` is rightly rejected, but `soma install codex --substrate claude-code --mode-classifier` depends on option order instead of the final parsed substrate.
  
  Suggestion:

  ```
  Validate `modeClassifier` support after parsing, against the final substrate value.
  ```

### Architecture
Architecture lens: the change stays within the Claude Code adapter and opt-in install surface, with the classifier behavior documented against the adapter contract. No new architecture findings; architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found).

_No findings._

### Performance
The PR adds an opt-in hook that runs on every Claude Code prompt submission. The only concrete performance concern is per-prompt CLI process startup and a long synchronous wait in the hook path.

- **[important]** `src/adapters/claude-code/mode-classifier-hook.mjs:31` — **Synchronous CLI spawn on every prompt**
  The UserPromptSubmit hook runs `spawnSync(config.bunPath, ["src/cli.ts", "algorithm", "classify", "--prompt", prompt || "", "--json"], { ... timeout: 25000 ... })`, which blocks prompt submission while starting a new Bun CLI process for every user prompt. That adds clearly inferable hot-path latency to each prompt and can stall Claude Code for up to 25 seconds before continuing.
  
  Suggestion:

  ```
  Use a lightweight classifier module in-process, or reduce the hook to a short timeout and fail open quickly.
  ```

### Maintainability
The change is mostly decomposed into focused helpers, but the Claude Code hook installer now repeats the same asset/config write pattern for each hook type. That duplication is small enough to be a suggestion, not a merge blocker.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:403` — **Deduplicate hook asset installation**
  The mode-classifier installer repeats the lifecycle hook installer shape: "await mkdir(dirname(hookPath), { recursive: true });", "await writeFile(hookPath, renderClaudeCodeModeClassifierHook(), { encoding: \"utf8\", mode: 0o755 });", "await chmod(hookPath, 0o755);", and "await writeFile(configPath, `${JSON.stringify(config, null, 2)}\n`, \"utf8\");". Adding more Claude Code hook assets would require copy-pasting this same sequence again.
  
  Suggestion:

  ```
  Extract a small installClaudeCodeHookAsset({ hookPath, configPath, source, config }) helper.
  ```


---
_Posted by Sage on Codex CLI substrate._

File: src/adapters/claude-code/mode-classifier-hook.mjs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    if (typeof input[key] === "string") return input[key];
+  }
+  return "";
+}
+
+function runSomaClassification(config, prompt) {
+  return spawnSync(config.bunPath, ["src/cli.ts", "algorithm", "classify", "--prompt", prompt || "", "--json"], {
+    cwd: config.trustedSomaRepo,
+    encoding: "utf8",
+    timeout: 25000,
```
Comment (jcfischer) [inline src/adapters/claude-code/mode-classifier-hook.mjs]: **[major]** Timeout mismatch: the settings entry registers this hook with `timeout: 10` (seconds), but `spawnSync` here uses `timeout: 25000` (25 seconds). Claude Code will kill the hook process at 10s, but the child `soma algorithm classify` keeps running for up to 25s as an orphan.

Either:
- Raise the settings timeout to ~30s to match, or
- Lower the `spawnSync` timeout to ~8s so it finishes before Claude Code kills the parent.

The second option is safer — a 25s synchronous hook on every prompt submission is a rough UX.

File: src/adapters/claude-code/mode-classifier-hook.mjs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    const parsed = JSON.parse(raw);
+    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : {};
+  } catch {
+    return {};
+  }
+}
+
+function promptFromInput(input) {
+  for (const key of ["prompt", "userPrompt", "message", "input"]) {
+    if (typeof input[key] === "string") return input[key];
```
Comment (jcfischer) [inline src/adapters/claude-code/mode-classifier-hook.mjs:34]: **[nit]** `promptFromInput` tries 4 key names (`prompt`, `userPrompt`, `message`, `input`). Claude Code's `UserPromptSubmit` hook contract only sends `prompt`. The extra fallbacks aren't wrong but could mask a schema change — if Claude Code stops sending `prompt`, silently falling back to `message` would produce confusing behavior rather than a clear error.

**Review by jcfischer — COMMENTED** [review]
## Code Review — soma#284 (Add opt-in Claude Code mode classifier)

**Lenses applied:** CodeQuality, Security, Architecture

**Overall:** Clean, well-structured PR. The opt-in pattern is right — mode classification as a flag, not default behavior. PAI hook disable/restore lifecycle is thorough with good idempotency (double-install tested). Tests cover the full install → classify → uninstall round-trip.

### Findings summary

| Severity | Count | Details |
|----------|-------|---------|
| Blockers | 0 | — |
| Majors | 1 | Hook timeout mismatch (settings vs spawnSync) |
| Nits | 3 | Regex redundancy, empty catch, minor defensive note |

See inline comments for details.

### Duplication check
No concerning duplication. The `removeCommandsFromSettingsEvent` extraction is a good DRY improvement — the existing `removeSomaHooksFromSettings` could delegate to it in a follow-up.

File: src/adapters/claude-code/hooks.ts [hunk 3]
```diff
@@ -176,6 +254,97 @@ function removeSomaCommandsFromGroup(group: unknown, commands: Set<string>): { g
   return { group: { ...group, hooks: nextHooks }, changed: true };
 }
 
+function isPaiModeClassifierCommand(command: string): boolean {
+  return /(?:^|[/\\])ModeClassifier\.hook\.(?:ts|js|mjs)(?:\s|$)/.test(command) || command.includes("ModeClassifier.hook.ts");
```
Comment (jcfischer) [inline src/adapters/claude-code/hooks.ts]: **[nit]** `isPaiModeClassifierCommand` has a redundant fallback: the regex already matches `ModeClassifier.hook.ts` (via the `(?:ts|js|mjs)` alternation), so the `|| command.includes("ModeClassifier.hook.ts")` branch can never add a match the regex missed. Not harmful, but misleading — suggests there is a case the regex does not cover.

File: src/adapters/claude-code/hooks.ts [hunk 4]
```diff
@@ -267,10 +481,32 @@ async function readInstalledClaudeCodeHookBunPath(substrateHome: string): Promis
...
+    throw error;
+  });
+  if (content === undefined) return undefined;
+  try {
+    const parsed: unknown = JSON.parse(content);
+    if (isObject(parsed) && typeof parsed.bunPath === "string" && parsed.bunPath.trim().length > 0) {
+      return parsed.bunPath;
+    }
+  } catch {
+    return undefined;
```
Comment (jcfischer) [inline src/adapters/claude-code/hooks.ts:515]: **[nit]** Empty `catch` block silently swallows JSON parse errors. Repo critical rule says every catch must log, handle, or name the variable `_err` with a comment. The existing `readInstalledClaudeCodeHookBunPath` has the same pattern (pre-existing), but since this is new code, worth fixing here.

**Review by jcfischer — COMMENTED** [review]
## Sage code review — changes-requested

2 finding(s): 1 important, 1 suggestion; 1 lens(es) failed to run: Performance.


### CodeQuality
The change is generally coherent for CodeQuality: option parsing, hook installation, settings mutation, and tests line up with the requested behavior. I found no new correctness or error-handling issues in this diff that were not already covered by previous Sage findings.

_No findings._

### Architecture
Architecture lens found no new contract-level issues in this diff. The Claude Code-specific hook work stays inside the Claude Code adapter surface, and previously raised concerns are not re-raised; architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found).

_No findings._

### Performance — DID NOT RUN
> ⚠ Lens failed to execute. Verdict cannot rely on this lens's coverage; re-run before merging.

- **[important]** `(lens output):0` — **Performance: model deviated from JSON contract**
  codex extraction failed (no-extractor-matched)
  - raw: undefined
  - fenced-last-first: undefined
  - balanced-largest: undefined
  - trailing: undefined
--- output ---
{"summary":"The main prompt-path performance cost, synchronous classification per prompt, was already raised and should not be duplicated here. I found one additional failure-path context growth issue that can add avoidable token latency when classification fails.","findings":[{"path":"src/adapters/claude-code/mode-classifier-hook.mjs","line":89,"severity":"suggestion","title":"Truncate classifier failure output","rationale":"The failure path injects raw command output via `additionalContext: \`Soma mode classifier failed; continue natively unless this prompt clearly needs Algorithm. ${result.stderr || result.stdout || \"\"}\``. If the classifier fails with a large stack trace or log, every prompt can carry unnecessary context tokens and latency.","suggestion":"Limit stderr/stdout to a small fixed prefix before adding it to `additionalContext`."}]}

### Maintainability
The diff is mostly maintainable after the hook-asset extraction, but one small duplication remains in config parsing. No maintainability issue rises above suggestion severity.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:497` — **Deduplicate bunPath config readers**
  `readInstalledClaudeCodeModeClassifierBunPath` repeats the same read/parse/validate shape as the existing hook reader: `const content = await readFile(configPath, "utf8").catch((error: unknown) => {` followed by `JSON.parse(content)` and `typeof parsed.bunPath === "string"`. Keeping two copies makes future config validation changes easy to miss in one path.
  
  Suggestion:

  ```
  Extract `readInstalledClaudeCodeHookConfigBunPath(relativePath)` and call it from both readers.
  ```


---
_Posted by Sage on Codex CLI substrate._

**Review by jcfischer — COMMENTED** [review]
## Sage code review — changes-requested

3 finding(s): 1 important, 1 suggestion, 1 nit.


### CodeQuality
The PR is generally coherent from a code-quality perspective, with focused tests covering the opt-in install and uninstall path. I found one edge-case correctness issue in the restore path that can report and write a settings change without actually restoring any valid hook.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:335` — **Avoid restoring empty hook groups**
  `hooks.UserPromptSubmit = groups;` runs even when every stored disabled entry is skipped because `commands.length === 0`. That can leave an empty `UserPromptSubmit` array and returns `true` despite restoring no hook.
  
  Suggestion:

  ```
  Track whether a group was appended, and only assign `hooks.UserPromptSubmit` / return `true` when restoration actually happened.
  ```

### Architecture
The change mostly keeps the classifier inside the Claude Code adapter, but it leaks a Claude-only install option into the shared install options type. Architecture docs loaded: architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found)

- **[important]** `src/types.ts:462` — **Claude option leaks into shared install type**
  The diff adds `modeClassifier?: boolean;` to the shared `SomaInstallOptions`, but docs/architecture.md says the core is `substrate-neutral` and that adapters own substrate-native primitives. This makes a Claude Code `UserPromptSubmit` hook capability part of the public cross-substrate install surface, so future substrate-specific install features either widen this shared type with more one-off flags or require a breaking reshaping later.
  
  Suggestion:

  ```
  Move this behind a Claude Code-specific install options shape or adapter-owned post-projection option.
  ```

### Performance
No new performance findings in the current diff beyond the previously raised synchronous classifier invocation on each prompt. The remaining changes are install-time settings edits or small classifier table additions without concrete hot-path cost.

_No findings._

### Maintainability
The diff is mostly decomposed into small helpers after the prior duplication feedback. I found one minor maintainability issue where constants are introduced but bypassed in the cleanup path, making future key renames easier to miss.

- **[nit]** `src/adapters/claude-code/hooks.ts:277` — **Use existing hook-key constants**
  The file defines `const SOMA_DISABLED_HOOKS_KEY = "somaDisabledHooks";` and `const PAI_MODE_CLASSIFIER_KEY = "paiModeClassifier";`, but cleanup uses literal property names: `delete disabled.paiModeClassifier;` and `delete settings.somaDisabledHooks;`. That creates a small duplicate source of truth for the same persisted keys.
  
  Suggestion:

  ```
  Use `delete disabled[PAI_MODE_CLASSIFIER_KEY]` and `delete settings[SOMA_DISABLED_HOOKS_KEY]`.
  ```


---
_Posted by Sage on Codex CLI substrate._

**Review by jcfischer — COMMENTED** [review]
## Sage code review — changes-requested

3 finding(s): 1 important, 2 suggestion.


### CodeQuality
The PR is generally coherent for the CodeQuality lens: install, uninstall, and hook behavior are covered by focused tests. I found one edge-case cleanup issue in the PAI hook restore path.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:346` — **Clear stale disabled hook state**
  When a disabled PAI hook is already present in settings, `if (commands.length === 0 || commands.every((command) => existingCommands.has(command))) continue;` skips it, then `if (!restored) return false;` leaves `somaDisabledHooks.paiModeClassifier` behind. That stale state can survive uninstall even though there is nothing left to restore.
  
  Suggestion:

  ```
  After processing disabled groups, clear the stored PAI entry when all disabled commands are already present.
  ```

### Architecture
The opt-in classifier mostly stays inside the Claude Code adapter, but the install planner now reaches into Claude hook internals for optional projection files. architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found)

- **[important]** `src/install.ts:10` — **Planner imports adapter hook internals**
  `src/install.ts` now imports `SOMA_CLAUDE_MODE_CLASSIFIER_CONFIG_RELATIVE_PATH` and `SOMA_CLAUDE_MODE_CLASSIFIER_RELATIVE_PATH` from `./adapters/claude-code/hooks`, then splices them into the generic install plan. This violates the adapter boundary in docs/architecture.md: “Adapters project Soma into substrate-native primitives,” because the shared install planner now knows Claude Code hook file layout.
  
  Suggestion:

  ```
  Move optional Claude Code substrate files into the Claude install spec or a Claude adapter planner helper, and have `install.ts` consume that spec output.
  ```

### Performance
The current diff does not add any new concrete performance regression beyond the already-raised synchronous classifier spawn on each prompt. Install/uninstall settings work is off the hot path and bounded by small Claude settings files.

_No findings._

### Maintainability
The diff is generally maintainable, with the main complexity contained in the Claude Code hook settings helpers. I found one extractable duplication pattern that would lower future change cost without blocking the PR.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:160` — **Extract hook group append logic**
  `appendSomaModeClassifierHookGroup` repeats the same scan-and-append shape as the existing hook append path: `const eventGroups = Array.isArray(hooks.UserPromptSubmit) ? hooks.UserPromptSubmit as unknown[] : [];`, `for (const group of eventGroups)`, and `eventGroups.push({ ... hooks: [...] })`. Keeping two variants means future changes to hook dedupe or group writing need to be made in both places.
  
  Suggestion:

  ```
  Extract a small `appendCommandHookGroup(settings, event, entry, knownCommands, description)` helper and use it for both append paths.
  ```


---
_Posted by Sage on Codex CLI substrate._

**Review by jcfischer — COMMENTED** [review]
## Sage code review — commented

3 finding(s): 3 suggestion.


### CodeQuality
The opt-in classifier path is mostly clear and covered by install/uninstall tests. I found two edge-case correctness gaps in hook restore and hook startup error handling.

- **[suggestion]** `src/adapters/claude-code/hooks.ts:338` — **Avoid partial restore duplicates**
  `if (commands.length === 0 || commands.every((command) => existingCommands.has(command))) continue; groups.push(group);` restores the whole disabled group when only some commands already exist. That can duplicate an existing UserPromptSubmit command on uninstall.
  
  Suggestion:

  ```
  Filter each stored group's hooks to commands not already in existingCommands before pushing it.
  ```

- **[suggestion]** `src/adapters/claude-code/mode-classifier-hook.mjs:12` — **Fail open on config errors**
  `return JSON.parse(readFileSync(join(hookDir(), "soma-mode-classifier.config.json"), "utf8"));` can throw before the hook emits a `continue: true` payload. A missing or corrupt config would make prompt submission see a hook failure instead of the intended fail-open behavior.
  
  Suggestion:

  ```
  Wrap config loading and emit a continue payload with fallback context when it fails.
  ```

### Architecture
The PR now keeps the Claude Code classifier behind the Claude Code adapter/install surface without introducing a new cross-substrate abstraction violation. architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found)

_No findings._

### Performance
The main prompt-path cost was already raised in the previous Performance pass. I do not see additional concrete performance regressions in the current diff that warrant a new finding.

_No findings._

### Maintainability
The diff is mostly maintainable after the earlier cleanup, with the new hook logic split into focused helpers. I found one remaining test-maintainability issue where a single integration test now spans several independent behaviors.

- **[suggestion]** `test/claude-code-install.test.ts:220` — **Split broad classifier lifecycle test**
  The test combines install idempotency, PAI hook disabling, hook execution, uninstall, and restore in one flow: "await installSomaForClaudeCode({ homeDir, modeClassifier: true });", "const output = runClaudeModeClassifierHook", and "const removed = await uninstallSomaForClaudeCode({ homeDir });". That makes future failures harder to localize than separate behavior-focused tests.
  
  Suggestion:

  ```
  Split into install/disable, hook-output, and uninstall/restore tests using a shared fixture helper.
  ```


---
_Posted by Sage on Codex CLI substrate._

**Review by jcfischer — COMMENTED** [review]
## Sage code review — changes-requested

1 finding(s): 1 important.


### CodeQuality
The diff is generally coherent for the CodeQuality lens, with the main flows covered by targeted tests. I found one concrete edge-case bug in hook detection that can leave a conflicting PAI classifier enabled.

- **[important]** `src/adapters/claude-code/hooks.ts:257` — **Quoted PAI hooks may stay enabled**
  The detector only matches `.js`/`.mjs` when the extension is followed by whitespace or end of string: `/(?:^|[/\\])ModeClassifier\.hook\.(?:ts|js|mjs)(?:\s|$)/.test(command) || command.includes("ModeClassifier.hook.ts")`. A valid command like `bun "/path/ModeClassifier.hook.mjs"` has a quote after `.mjs`, so it is not disabled and Soma's classifier can run alongside it.
  
  Suggestion:

  ```
  Allow quote characters after the extension, e.g. `(?:[\s"']|$)`, and add a quoted `.mjs` test.
  ```

### Architecture
Architecture review finds no new blocking boundary or public-surface issues beyond the previously raised items. The diff keeps the classifier opt-in inside the Claude Code adapter surface and documents it as substrate-native hook behavior. architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found)

_No findings._

### Performance
The only concrete hot-path performance risk in this diff is the per-prompt synchronous classifier invocation, and that was already raised previously. The remaining install/uninstall settings mutations are bounded and off the prompt path, so this lens has no new findings.

_No findings._

### Maintainability
The diff is reasonably structured for this lens; the larger hook-settings logic has been split into targeted helpers and prior duplication concerns appear addressed. I found no new maintainability issues beyond the previously raised items.

_No findings._


---
_Posted by Sage on Codex CLI substrate._

**Review by jcfischer — COMMENTED** [review]
## Sage code review — approved

No findings. Sage approves.


### CodeQuality
The CodeQuality pass found no new correctness or edge-case issues in the current diff beyond the previously raised findings. The implementation includes targeted tests for opt-in install, uninstall restoration, quoted PAI commands, and fail-open behavior.

_No findings._

### Architecture
The architecture-relevant changes keep the Claude Code-specific hook behavior inside the Claude Code adapter and use the documented Algorithm classifier surface. architecture-docs: CONTEXT.md (loaded), docs/architecture.md (loaded), compass/ecosystem/CONTEXT-MAP.md (not-found)

_No findings._

### Performance
The diff adds opt-in prompt-time classification, but the concrete hot-path cost already appears in previously raised findings. I found no additional Performance-lens issues in the current diff that meet the grounding and severity bar.

_No findings._

### Maintainability
The diff is mostly maintainable: the hook mutation logic has been decomposed into focused helpers, and the asset install duplication noted earlier is addressed. I do not see new maintainability-only issues that were not already raised in prior review rounds.

_No findings._


---
_Posted by Sage on Codex CLI substrate._
